### PR TITLE
Mcs manifest

### PR DESCRIPTION
**Pull request by aravindakidambi**
Change manifest generation logic to generate
1. Member cluster yaml
2. Leader cluster global yaml - which needs to be applied once in the
   leader cluster
3. Leader namespaced yaml - which needs to be applied once per cluster-set
   namespace.
   e.g. to apply leader namespaced yaml
   $ kubectl create ns pepsi
   $ hack/generate-manifest.sh -l pepsi | kubectl apply -f -

### CODE REVIEW

File: multicluster/config/overlays/leader-global/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+namespace: coke
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-global/kustomization.yaml]: why it is coke here?

File: multicluster/config/overlays/leader-ns/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+namespace: coke
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-ns/kustomization.yaml]: ditto

File: multicluster/config/overlays/leader-ns/prefix_transformer.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+apiVersion: builtin
+kind: PrefixSuffixTransformer
+metadata:
+  name: customPrefixer
+prefix: coke-
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-ns/prefix_transformer.yaml]: ditto

File: multicluster/hack/generate-manifest.sh [hunk 4]
```diff
@@ -20,9 +20,10 @@ function echoerr {
     >&2 echo "$@"
 }
 
-_usage="Usage: $0 [--leader] [--member] [--help|-h]
+_usage="Usage: $0 [--global|-g] [--leader|-l] [--member|-m] [--help|-h]
 Generate a YAML manifest for Antrea MultiCluster using Kustomize and print it to stdout.
-        --leader                       Generate a manifest for a Cluster as leader in a ClusterSet
+        --global                       Generate a global manifest for a Cluster as leader in a ClusterSet
+        --leader		       Generate a per-namespace manifest for a Cluster as leader in a ClusterSet
```
Comment (luolanzone) [inline multicluster/hack/generate-manifest.sh]: make the ‘--leader’ line align with above one.

File: multicluster/config/overlays/leader-ns/webhook_patch.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+apiVersion: admissionregistration.k8s.io/v1
+kind: MutatingWebhookConfiguration
+metadata:
+  name: mutating-webhook-configuration
+webhooks:
+- name: mclusterclaim.kb.io
+  namespaceSelector:
+    matchLabels:
+      kubernetes.io/metadata.name: coke
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-ns/webhook_patch.yaml]: ditto, if we are giving fixed namespace for leader, we probably need to choose a proper one, or allow user to pass the namespace from command line

File: multicluster/config/rbac/kustomization.yaml [hunk 6]
```diff
@@ -9,9 +9,9 @@ resources:
 - role_binding.yaml
 - leader_election_role.yaml
 - leader_election_role_binding.yaml
-- secret_viewer_role.yaml
-- secret_viewer_role_binding.yaml
-- memberclusterannounce_editor_role.yaml
+# - secret_viewer_role.yaml
```
Comment (luolanzone) [inline multicluster/config/rbac/kustomization.yaml]: I think you can just remove it if it's not used here.

File: multicluster/config/webhook/kustomization.yaml [hunk 7]
```diff
@@ -1,6 +1,12 @@
+#namePrefix: antrea-multicluster-
+#
+#commonlabels:
+#  app: antrea
+
```
Comment (luolanzone) [inline multicluster/config/webhook/kustomization.yaml]: ditto

File: multicluster/build/yamls/antrea-multicluster-leader-ns.yml [hunk 8]
```diff
@@ -0,0 +1,722 @@
+apiVersion: v1
```
Comment (luolanzone) [inline multicluster/build/yamls/antrea-multicluster-leader-ns.yml]: I feel the file name `antrea-multicluster-leader-ns.yml` is unclear, maybe `antrea-multicluster-leader-namespaced.yml`? or you can think out a better name.

**Comment by luolanzone** [comment]
@aravindakidambi  Could you help also to add a simple doc about how to use this manifests? eg: how to replace the namespace and how to use those manifests, I feel it's also necessary to point out the service-account and secret creation setup for ClusterSet initialization.

File: multicluster/config/overlays/leader-ns/webhook_patch.yaml (hunk 5 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/webhook_patch.yaml]: can you tell me what would be a proper one ? I also tried to pass the namespace in command line, but Im not able to set the namespace correctly in all places esp in this matchLabels. So for now, it needs to be done manually after manifest is generated. Ill explore more and ask help and see if there is a way to improve it. But for now, kindly tell me what is a proper namespace name and I can use it.

**Comment by aravindakidambi** [comment]
> @aravindakidambi Could you help also to add a simple doc about how to use this manifests? eg: how to replace the namespace and how to use those manifests, I feel it's also necessary to point out the service-account and secret creation setup for ClusterSet initialization.

Sure, for now, I added how to use the leader-namespaced.yml in the commit description. I see you already have a change in flight to add docs: https://github.com/antrea-io/antrea/pull/2844/files

Can you merge it and I will append to that?

File: multicluster/config/overlays/leader-global/kustomization.yaml (hunk 1 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-global/kustomization.yaml]: changing to changeme per Akshay's suggestion

File: multicluster/config/overlays/leader-ns/kustomization.yaml (hunk 2 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/kustomization.yaml]: changing to changeme per Akshay's suggestion

File: multicluster/config/overlays/leader-ns/prefix_transformer.yaml (hunk 3 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/prefix_transformer.yaml]: changing to changeme per Akshay's suggestion

File: multicluster/config/overlays/leader-ns/webhook_patch.yaml (hunk 5 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/webhook_patch.yaml]: changing to changeme per Akshay's suggestion

File: multicluster/cmd/multicluster-controller/controller.go [hunk 9]
```diff
@@ -64,6 +64,19 @@ func init() {
 	//+kubebuilder:scaffold:scheme
 }
 
+func getValidationWebhooks(isLeader bool) []string {
+	if isLeader {
+		return []string{fmt.Sprintf(validationWebhooksNamePattern, env.GetPodNamespace(), "-")}
+	}
+	return []string{fmt.Sprintf(validationWebhooksNamePattern, "", "")}
+}
+func getMutationWebhooks(isLeader bool) []string {
```
Comment (abhiraut) [inline multicluster/cmd/multicluster-controller/controller.go:74]: missing line

File: multicluster/config/overlays/leader-global/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+namespace: changeme
```
Comment (abhiraut) [inline multicluster/config/overlays/leader-global/kustomization.yaml]: could leave a comment above describing what name change is expected

File: multicluster/config/overlays/leader-ns/webhook_rbac.yaml [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    app: antrea
+  name: controller-webhook-rolebinding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: controller-webhook-role
+subjects:
+  - kind: ServiceAccount
+    name: controller
+    namespace: system
```
Comment (abhiraut) [inline multicluster/config/overlays/leader-ns/webhook_rbac.yaml]: system ns?

Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/webhook_rbac.yaml]: It will get overwritten, I just copied whatever was there previously.

Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/webhook_rbac.yaml]: I changed it to changeme and also fixed the manifest generation to take namespace name as input

File: multicluster/config/overlays/leader-global/kustomization.yaml (hunk 10 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-global/kustomization.yaml]: None of these are actually needed for leader-global, so I removed them.

File: multicluster/config/overlays/leader-ns/kustomizeconfig.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
+nameReference:
+  # Since mutating webhook configuration name gets prefixed during kustomization,
+  # propagate the actual names into the corresponding cluster-role that is
+  # necessary for mcs-controller to have access to those webhook configuration.
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-ns/kustomizeconfig.yaml]: s/configuration/configurations/

File: multicluster/config/overlays/leader-ns/kustomizeconfig.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+- kind: MutatingWebhookConfiguration
+  version: v1
+  fieldSpecs:
+    - kind: ClusterRole
+      group: rbac.authorization.k8s.io
+      path: rules/resourceNames
+  # Since validating webhook configuration name gets prefixed during kustomization,
+  # propagate the actual names into the corresponding cluster-role that is
+  # necessary for mcs-controller to have access to those webhook configuration.
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-ns/kustomizeconfig.yaml]: ditto

File: multicluster/config/overlays/leader-ns/member_cluster_rbac.yaml [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      - list
+      - watch
+  - apiGroups:
+      - multicluster.crd.antrea.io
+    resources:
+      - resourceimports/status
+    verbs:
+      - get
+      - patch
+      - update
```
Comment (luolanzone) [inline multicluster/config/overlays/leader-ns/member_cluster_rbac.yaml]: missing new line.

File: multicluster/config/overlays/member/kustomization.yaml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+commonLabels:
+  app: antrea
+
+bases:
+  - ../../crd
+  - ../../default
+
+resources:
+  - secret_viewer_role.yaml
+  - secret_viewer_role_binding.yaml
```
Comment (luolanzone) [inline multicluster/config/overlays/member/kustomization.yaml]: do we still need this standalone secret view and role binding? I think secret view role has been added via `additional_role_patch.yaml`

File: multicluster/config/overlays/leader-ns/kustomizeconfig.yaml (hunk 12 above)
Comment (luolanzone) [reply multicluster/config/overlays/leader-ns/kustomizeconfig.yaml]: btw, the description in issue summary needs update.

File: multicluster/config/overlays/member/kustomization.yaml (hunk 15 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/member/kustomization.yaml]: You are right, its not needed.

File: multicluster/config/overlays/leader-ns/kustomizeconfig.yaml (hunk 12 above)
Comment (aravindakidambi) [reply multicluster/config/overlays/leader-ns/kustomizeconfig.yaml]: I did update commit message, dont know why it doesnt show up. Ill try to look into it

File: multicluster/config/overlays/member/kustomization.yaml [hunk 16]
```diff
@@ -0,0 +1,27 @@
+namespace: kube-system
```
Comment (luolanzone) [inline multicluster/config/overlays/member/kustomization.yaml:1]: should we also allow user to update this namespace? this is configured via `ClusterSet`, right?

Comment (aravindakidambi) [reply multicluster/config/overlays/member/kustomization.yaml:1]: for member cluster, it doesnt matter because it has cluster-wide access to resources.

**Pull request merged**